### ISSUE TITLE
No MIR optimisations with no-trans

### ISSUE THREAD

**Issue by leonardo-m** [rust-lang/rust#41423]
This is an enhancement request. When I compile the code with "-Z no-trans" to just verify the types and perform the borrow checking, I have seen with "-Ztime-passes" that it also performs (on my code, a large single file module):

`time: 0.127; rss: 177MB MIR optimisations`

Is it possible and a good idea for "-Z no-trans" to remove this part too? The gain is not a lot (about 0.13 seconds on about 3.5 seconds) but if this optimization is simple to implement, I think it's worth it, because I'm using no-trans often enough to test the code.

**Comment by eddyb** [comment]
cc @rust-lang/compiler Should we do this? Not by treating MIR optimizations as part of "trans" but by not optimizing the MIR that doesn't end up in the crate metadata.

**Comment by nikomatsakis** [comment]
@eddyb I would think that we want the optimized MIR in the metadata, since if we have to inline across crates, we don't want to optimize *again*, right?

**Comment by eddyb** [comment]
@nikomatsakis Right, but we only put MIR in the metadata if it's needed across crates.
We could even do some reachability analysis to reduce that number.

**Comment by nagisa** [comment]
Two points. I believe `-Z no-trans` is not the way to do the "check". It is `--emit=metadata` now. And for `--emit=metadata` it is what @nikomatsakis said.

I feel like this is one of the things that could be automatically solved as we make more stuff on-demand.

**Comment by arielb1** [comment]
On-demand compilation could significantly change the story here. Let's wait for it.

**Comment by nikomatsakis** [comment]
triage: P-low 

We are still working out the story around mir optimizations and the query system. Once that is more sorted, we can revisit this question. We're barely doing any MIR optimization now anyhow.

**Comment by steveklabnik** [comment]
Triage: no changes that I'm aware of

**Comment by cjgillot** [comment]
Check builds do not optimize MIR any more.
